File: Data Structure and Algorithm/Assignment 03/Program 6/prog6.c
// Write a program to print the odd number and even number nodes separately from a
// single linked list.


#include<stdio.h>
#include<stdlib.h>

typedef struct Node{
    int data;
    struct Node * next;
}node;

node *createNode_at_first(){
    node *nw, *head = NULL;
    int ch = 1,data;
    while (ch!=0)
    {
        printf("Enter the data: ");
        scanf("%d",&data);

        nw = (node*)malloc(sizeof(node));
        nw->data = data;
        nw->next = head;
        head = nw;

        printf("Do you want to input:");
        scanf("%d",&ch);
    }
    return head;
}

node *deleteNode_at_first(node *head){
    node *p =head;
    head = head->next;
    free(p);
    return head;
}

node *createNode_at_last(node *head){
    node *nw, *p;
    int ch = 1,data;
    while (ch!=0)
    {
        printf("Enter the data: ");
        scanf("%d",&data);

        nw = (node*)malloc(sizeof(node));
        nw->data = data;
        nw->next = NULL;
        if (head == NULL)
            head = nw;
        else{
            p = head;
            while (p->next!= NULL)
                p = p->next;
            p->next = nw;
            nw->next = NULL;
        }

        printf("Do you want to input:");
        scanf("%d",&ch);
    }
    return head;
}

node *deleteNode_at_last(node *head){
    if (head == NULL){
        printf("Empty list");
        return NULL;
    }
    else if(head->next == NULL){
        free(head);
    }
    else{
        node *p =head;
        node *q = p->next;
        while (q->next!= NULL)
        {
            p = p->next;
            q = q->next;
        }
        p->next = NULL;
        free(q);
}
    return head;
}

node *createNode_at_index(node *head, int index){
    node *nw, *p;
    int ch = 1,i=0,data;
    while (ch!=0)
    {
        printf("Enter the data: ");
        scanf("%d",&data);

        nw = (node*)malloc(sizeof(node));
        nw->data = data;
        nw->next = NULL;
        
        if (index == 0){
            nw->next = head;
            head = nw;
            return head;
        }
        else{
            p = head;
            while (p!=NULL && i < index-1)
            {
                p = p->next;
                i++;
            }
            nw->next = p->next;
            p->next = nw;
        }

        printf("Do you want to input:");
        scanf("%d",&ch);
    }
    return head;
}

node *deleteNode_at_index(node *head,int index){
     if (head == NULL) {
        printf("List is empty.\n");
        return NULL;
    }
    else{
        node *p = head;
        int i = 0;
        if(index == 0){
            node *temp = head;
            head = head->next;
            free(temp);
            return head;
        }
        while(p != NULL && i<index-1){
            p = p->next;
            i++;
        }
        node *q = p->next;
        p->next = q->next;
        free(q);
    }
    return head;
}

void free_list(node* head) {
    node* temp;
    while (head != NULL) {
        temp = head;
        head = head->next;
        free(temp);
    }
}

void display(node *head){
    node *temp = head;
    while(temp!= NULL){
        printf("%d ->",temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

void print_even_odd(node *head){
    if(head == NULL){
        printf("Empty list");
        return ;
    }
    else{
        node *p = head;
        while (p!= NULL){
            if(p->data%2 == 0){
                printf("Even : %d -->",p->data);
            }
            p = p->next;
        }
        printf("NULL");
        printf("\n");
        node *q = head;
        while (q!= NULL){
            if(q->data%2 != 0){
                printf("Odd : %d -->",q->data);
            }
            q = q->next;
        }
        printf("NULL");
    }
}

int main(){
     node *head;
    int ch,index;
    do{
        printf("\n____MENU____\nEnter your choice\n1.Insert at first\n2.Insert at last\n3.Insert at index\n4.Delete at first\n5.Delete at last\n6.Delete at index\n7.Print even odd seperately\n8.Exit\n\n");
        scanf("%d",&ch);

        switch(ch){
            case 1:
                head = createNode_at_first();
                display(head);
                break;
            case 2:
                head = createNode_at_last(head);
                display(head);
                break;
            case 3:
                printf("Enter the index no.: ");
                scanf("%d",&index);
                head = createNode_at_index(head,index);
                display(head);
                break;
            case 4:
                head = deleteNode_at_first(head);
                display(head);
                break;
            case 5:
                head = deleteNode_at_last(head);
                display(head);
                break;
            case 6:
                printf("Enter the index no.: ");
                scanf("%d",&index);
                head = deleteNode_at_index(head,index);
                display(head);
                break;

            case 7:
                print_even_odd(head);
                break;
            case 8:
                printf("Exiting...");
                break;
            default:
                printf("You give wrong input");
                break;
        }
    }while(ch != 8);

    free_list(head);
    return 0;
}

File: Data Structure and Algorithm/Assignment 03/Program 6/prog6.py
# Write a program to print the odd number and even number nodes separately from a
# single linked list.


class node:
    def __init__(self,data):
        self.data =data
        self.next = None

class linkedList:
    def __init__(self):
        self.head = None

    def insert_at_first(self,data):
        new_node = node(data)
        new_node.next = self.head
        self.head = new_node

    def delete_at_first(self):
        if self.head is not None:
            self.head = self.head.next

    def insert_at_last(self,data):
        new_node = node(data)
        if self.head is None:
            self.head = new_node
            return 
        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node
        new_node.next = None

    def delete_at_last(self):
        if self.head == None:
            print("Empty List")
        if self.head.next == None:
            self.head = None
            return
        temp = self.head
        while temp.next.next:
            temp = temp.next
        temp.next = None
        
    def insert_at_index(self,data,index):
        new_node = node(data)
        i=0
        p = self.head
        if (index == 0):
            new_node.next = self.head
            self.head = new_node
            return
        while p and i< index-1:
            p = p.next
            i+=1

        new_node.next = p.next
        p.next = new_node

    def delete_at_index(self,index):
            if self.head == None:
                print("Empty List")
            i=0
            if index == 0:
                t = self.head
                self.head = self.head.next
                del(t)
                return
            temp = self.head
            while temp and i< index-1:
                temp = temp.next
                i+=1
            temp2 = temp.next
            temp.next = temp2.next
            del(temp2)
        
    def display(self):
        temp = self.head
        while temp:
            print(temp.data,end="-->")
            temp = temp.next
        print("None")

    def print_even_odd(self):
        if self.head == None:
            print("Empty List")
            return 
        else:
            p = self.head 
            print("Even: ")
            while p:
                if p.data%2 == 0:
                    print(p.data,end="-->")
                p = p.next
            print("None\n")

            q = self.head 
            print("Odd: ")
            while q:
                if q.data%2 != 0:
                    print(q.data,end="-->")
                q= q.next
            print("None\n")

if __name__ == "__main__":
    n = linkedList()
    ch = 0
    while ch != 8:
        ch = int(input("___MENU___\nEnter the choice\n1.CreateNode at first\n2.CreateNode at last\n3.CreateNode at index\n4.DeleteNode at first\n5.DeleteNode at last\n6.DeleteNode at index\n7.Print Even and odd\n\n"))
        match ch:
            case 1:
                data = int(input("Enter the data: "))
                n.insert_at_first(data)
                n.display()
            case 2:
                data = int(input("Enter the data: "))
                n.insert_at_last(data)
                n.display()
            case 3:
                index = int(input("Enter the index: "))
                data = int(input("Enter the data: "))
                n.insert_at_index(data,index)
                n.display()
            case 4:
                n.delete_at_first()
                n.display()
            case 5:
                n.delete_at_last()
                n.display()
            case 6:
                index = int(input("Enter the index: "))
                n.delete_at_index(index)
                n.display()
            case 7:
                n.print_even_odd()
            case 8:
                print("Exiting...")
        